docs/Grammar.md

```diff
@@ -0,0 +1,151 @@
+- `identifier` A sequence of alphanumeric characters or '_' that starts with a character or '_'.
+   Valid identifiers in EDL are valid C language identifiers.
+- `integer` A sequence of digits. EDL does not support octal and hexadecimal integers.
+
+Each `non-termial` is defined via `production` of the form
+```c
+     non-termial = expression
+```
+where `expression` consists of one or more `non-terminal`s or tokens.
+
+
+`+` Indicates one or more occurance of the preceeding symbol.
+
+`*` Indicates zero or more occurance of the preceesing symbol.
+
+`[]` Indicates that the enclosed item is optional.
+
+`()` Paranthesis is used to group items.
+
+## Grammar
+
+```c
+
+enclave_definition :  "enclave" '{' enclave_items* '}' EOF
+
+enclave_item :
+    include_statement
+    | import_statement
+    | from_import_statement
+    | enum_declaration
+    | struct_declaration
+    | union_declaration
+    | trusted_section
+    | untrusted_section
+
+include_statement: "include" string_token
+
+import_statement: "import" string_token
+
+from_import_statement: "from" string_token "import" imported_items ';'
+
+imported_items =
+    '*'
+    | identifier_list
+
+identifier_list = identifier ( ',' identifier )*
+
+enum_declaration =
+    "enum" [identifier] '{' enumeration_item (',' enumeration_item)* '}' ';'
+
+enumeration_item =
+    | identifier = integer | identifier
+    | identifier
+
+struct_declaration = "struct" '{' (declaration ';') * '}' ';'
+
+union_declaration = "union" '{' (declaration ';') * '}' ';'
+
+declaration = [attributes] type_expr identifier [array_dimension+]
+
+attributes = '[' attribute (',' attribute)* ']'
+
+attribute =
+    "in" | "out" | "string" | "wstring" | "isptr" | "isary" | "user_check"
+    | "count" = attribute_value
+    | "size" = attribute_value
+
+attribute_value = integer | identifier
+
+type_expr = [atype] [pointer]
+
+atype = ["const"] (atype1 | atype2)
+
+atype1 = "unsigned" integer_type
+    | integer_type
+    | "long" "double"
+
+integer_type =
+    | "long" "long" ["int"]
+    | "long" ["int"]
+    | "short" ["int"]
+    | "char" ["int"]
+    | "int"
+
+atype2 =
+    | "enum" identifier
+    | "struct" identifier
+    | "union" identifier
+    | "bool"
+    | "void"
+    | "wchar_t"
+    | "size_t"
+    | "int8_t"
+    | "int16_t"
+    | "int32_t"
+    | "int64_t"
+    | "uint8_t"
+    | "uint16_t"
+    | "uint32_t"
+    | "uint64_t"
+    | "float"
+
+pointer =
+    '*'
+    | '*' pointer
+
+array_dimension = '[' (integer | identifier) ']'
+
+trusted_section = "trusted" '{' trusted_function* '}' ';'
+
+trusted_function =
+    "public" function_declaration [allow_list] [trusted_suffixes] ';'
+
+allow_list = "allow" '(' identifier_list ')'
+
+untrusted_section =
+    "untrusted" '{' untrusted_function* '}' [untrusted_suffixes] ';'
+
+function_declaration = atype identifier parameter_list
+
+parameter_list =
+    '(' "void" ')'
+    | '(' ')'
+    | '(' declaration  ( ',' declaration )* ')'
```

Comment:
should there be ';' at the end?

Reply:
So the semi-colon is added at the end of `trusted_function` and `untrusted_function` productions.
This is because,
- after the parameter list, untrusted functions can have `transition_using_threads` and `propagate_errno` suffixes and then a semi-colon.
- after parameter list, trusted functions can have an allow list (which doesn't apply for OE since reentrant ecalls are not supported, and hence OE warns about allow lists), as well as `transition_using_threads` suffix.